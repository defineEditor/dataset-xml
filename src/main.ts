import fs from 'fs';
import flow from 'xml-flow';

export const read = async (pathToFile): Promise<object> => {
    return await new Promise((resolve, reject) => {
        const xmlFile = fs.createReadStream(pathToFile);
        const xmlStream = flow(xmlFile, { strict: true });

        let header = {};
        let lastGroupOID;
        const itemGroupData = {};

        xmlStream.on('tag:ItemGroupData', (xmlRecord) => {
            const itemGroupOID = xmlRecord.$attrs.ItemGroupOID;
            if (itemGroupOID !== lastGroupOID) {
                lastGroupOID = itemGroupOID;
                itemGroupData[itemGroupOID] = {
                    ItemDataOID: [],
                    ItemData: [],
                };
                xmlRecord.ItemData.forEach(itemData => {
                    itemGroupData[itemGroupOID].ItemDataOID.push(itemData.ItemOID);
                });
            }
            const record = [];
            xmlRecord.ItemData.forEach(itemData => {
                record.push(itemData.Value);
            });
            itemGroupData[itemGroupOID].ItemData.push(record);
        });

        xmlStream.on('tag:ReferenceData', (xmlRecord) => {
            header = {
                StudyOID: xmlRecord.$attrs.StudyOID,
                MetaDataVersionOID: xmlRecord.$attrs.MetaDataVersionOID,
            };
        });

        xmlStream.on('tag:ClinicalData', (xmlRecord) => {
            header = {
                StudyOID: xmlRecord.$attrs.StudyOID,
                MetaDataVersionOID: xmlRecord.$attrs.MetaDataVersionOID,
            };
        });

        xmlStream.on('end', () => {
            resolve({
                ...header,
                ItemGroupData: itemGroupData,
            });
        });
    });
};
