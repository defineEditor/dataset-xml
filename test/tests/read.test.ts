import path from 'path';
import { read } from '../../src/main';

describe('Read file', () => {
    it('Read dataset with ReferenceData', async () => {
        let result = await read(path.join(__dirname, '../data/TV.xml'));
        expect(result).toMatchSnapshot();
    });
    it('Read dataset with ClinicalData', async () => {
        let result = await read(path.join(__dirname, '../data/CO.xml'));
        expect(result).toMatchSnapshot();
    });
});