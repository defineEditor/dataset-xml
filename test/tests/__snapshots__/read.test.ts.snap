// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Read file Read dataset with ClinicalData 1`] = `
Object {
  "ItemGroupData": Object {
    "CO": Object {
      "ItemData": Array [
        Array [
          "CDISCPILOT01",
          "CO",
          "AE",
          "01-701-1023",
          "1",
          "AESEQ",
          "2",
          undefined,
          "related to record 1 in domain DS",
          "PRINCIPAL INVESTIGATOR",
        ],
        Array [
          "CDISCPILOT01",
          "CO",
          "AE",
          "01-701-1047",
          "2",
          "AESEQ",
          "4",
          undefined,
          "related to record 1 in domain DS. The quick brown fox jumps over the lazy dog and can easily have more than twohundred characters which was the upper limit that SAS Transport 5 could be dealing with. Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum",
          "PRINCIPAL INVESTIGATOR",
        ],
      ],
      "ItemDataOID": Array [
        "CO.STUDYID",
        "CO.DOMAIN",
        "CO.RDOMAIN",
        "CO.USUBJID",
        "CO.COSEQ",
        "CO.IDVAR",
        "CO.IDVARVAL",
        "CO.COREF",
        "CO.COVAL",
        "CO.COEVAL",
      ],
    },
  },
  "MetaDataVersionOID": "CDISC.SDTM.3.1.0",
  "StudyOID": "LZZT",
}
`;

exports[`Read file Read dataset with ReferenceData 1`] = `
Object {
  "ItemGroupData": Object {
    "TV": Object {
      "ItemData": Array [
        Array [
          "CDISCPILOT01",
          "TV",
          "1",
          "SCREENING 1",
          "-7",
          "Start of Screen Epoch",
          "completion of screening activities",
        ],
        Array [
          "CDISCPILOT01",
          "TV",
          "2",
          "SCREENING 2",
          "-1",
          "Start of Visitnum 3 - 24H",
          "subject leaves clinic after connection of ambulatory ECG machine",
        ],
        Array [
          "CDISCPILOT01",
          "TV",
          "3",
          "BASELINE",
          "1",
          "(Just Before) Start of Treatment Epoch",
          "After application of first dose",
        ],
      ],
      "ItemDataOID": Array [
        "TV.STUDYID",
        "TV.DOMAIN",
        "TV.VISITNUM",
        "TV.VISIT",
        "TV.VISITDY",
        "TV.TVSTRL",
        "TV.TVENRL",
      ],
    },
  },
  "MetaDataVersionOID": "CDISC.SDTM.3.1.0",
  "StudyOID": "LZZT",
}
`;
